docs/build/guides/testing/detecting-changes-with-test-snapshots.mdx

```diff
@@ -0,0 +1,51 @@
+---
+title: Detecting Unexpected Changes with Test Snapshots
+hide_table_of_contents: true
+description: Use test snapshots to detect unexpected changes in contract behavior
+---
+
+Tests are written to ensure that contracts behave today as expected, and in the future as well. Over time a contract may change and in all software development there remains the possibility of changes causing side-effects that are unexpected. Testing is one of the ways that we identify unexpected changes.
+
+However tests are limited, as they only show changes to values that the tests assert on.
+
+## Test Snapshots
+
+The Soroban Rust SDK generates test snapshots on every test involving an `Env`. Test snapshots are enabled by default and for every test `Env`. At the end of the test the `Env` writes a JSON file to the `test_snapshots` directory with a full snapshot of all the events published, logged, and the final ledger storage state.
+
+### How to Use Test Snapshots
+
+1. Write tests using the default `Env`. For example:
+
+   ```rust
+   #![cfg(test)]
+   use soroban_sdk::Env;
+
+   use crate::{Contract, ContractClient};
+
+   #[test]
+   fn test_abc() {
+       let env = Env::default();
+       let contract_id = env.register_contract(None, Contract);
+       let client = ContractClient::new(&env, &contract_id);
+
+       assert_eq!(client.increment(), 1);
+
+       // At the end of the test the Env will automatically write a test snapshot
+       // to the following directory: test_snapshots/test_abc.1.json
```

Comment:
Why are tests incremented? `test_abc.1.json` (why `.1.`?)

Reply:
What happens when you update tests or rename tests or do a refactor of your code? I assume test file names come from the function names? So once you write a test is it true you should not be renaming or changing your tests? You should just write new ones. 🤔

Reply:
> Why are tests incremented? test_abc.1.json (why .1.?)

Added some more documentation answering the question: b437c34e.

> What happens when you update tests or rename tests or do a refactor of your code?

New files will be written. Depending on the amount that changes git will probably identify it as a rename even if some parts of it change if the original file is removed. The handling of this isn't perfect but impact is small and not different to how some golden file or vcr frameworks will save files.